Userland/Libraries/LibUnicode/Locale.cpp

```diff
@@ -0,0 +1,205 @@
+        language_id.script = segments[index];
+        if (segments.size() == ++index)
+            return language_id;
+    } else if (!language_id.language.has_value()) {
+        return {};
+    }
+
+    if (is_unicode_region_subtag(segments[index])) {
+        language_id.region = segments[index];
+        if (segments.size() == ++index)
+            return language_id;
+    }
+
+    while (index < segments.size()) {
+        if (!is_unicode_variant_subtag(segments[index]))
+            return {};
+        language_id.variants.append(segments[index++]);
+    }
+
+    return language_id;
+}
+
+Optional<LocaleID> parse_unicode_locale_id(StringView locale)
+{
+    LocaleID locale_id {};
+
+    // https://unicode.org/reports/tr35/#Unicode_locale_identifier
+    // unicode_locale_id = unicode_language_id
+    //                     extensions*
+    //                     pu_extensions? ;
+    auto language_id = parse_unicode_language_id(locale);
+    if (!language_id.has_value())
+        return {};
+
+    // FIXME: Handle extensions and pu_extensions.
+    return LocaleID { language_id.release_value() };
+}
+
+Optional<String> canonicalize_unicode_locale_id(LocaleID& locale_id)
+{
+    // https://unicode.org/reports/tr35/#Canonical_Unicode_Locale_Identifiers
+    StringBuilder builder;
+
+    if (!locale_id.language_id.language.has_value())
+        return {};
+
+    builder.append(locale_id.language_id.language->to_lowercase_string());
+
+    if (locale_id.language_id.script.has_value()) {
+        builder.append('-');
+        builder.append(to_titlecase(*locale_id.language_id.script));
+    }
+
+    if (locale_id.language_id.region.has_value()) {
+        builder.append('-');
+        builder.append(locale_id.language_id.region->to_uppercase_string());
+    }
+
+    quick_sort(locale_id.language_id.variants);
+
+    for (auto const& variant : locale_id.language_id.variants) {
+        builder.append('-');
+        builder.append(variant.to_lowercase_string());
+    }
+
+    // FIXME: Handle extensions and pu_extensions.
+
+    return builder.build();
+}
+
+String const& default_locale()
+{
+    static String locale = "en"sv;
+    return locale;
+}
+
+bool is_locale_available(StringView locale)
+{
+#if ENABLE_UNICODE_DATA
+    static auto const& available_locales = Detail::available_locales();
+    return available_locales.contains(locale);
+#else
+    return false;
+#endif
+}
+
+Optional<StringView> get_locale_territory_mapping(StringView locale, StringView code)
+{
+#if ENABLE_UNICODE_DATA
+    static auto const& available_locales = Detail::available_locales();
+
+    auto it = available_locales.find(locale);
+    if (it == available_locales.end())
+        return {};
+
+    if (auto territory = Detail::territory_from_string(code); territory.has_value())
+        return it->value.territories[static_cast<TerritoryUnderlyingType>(*territory)];
+
+    return {};
+#else
+    return {};
+#endif
```

Comment:
(weak suggestion, just a style thing)
```suggestion

#endif 
    return {};
```